In [541]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('../input/heart-failure-clinical-data'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

../input/heart-failure-clinical-data/heart_failure_clinical_records_dataset.csv


In [542]:
clinical_records = pd.read_csv('../input/heart-failure-clinical-data/heart_failure_clinical_records_dataset.csv')

In [543]:
clinical_records.head()

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1

In [544]:
clinical_records.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       299 non-null    float64
 1   anaemia                   299 non-null    int64  
 2   creatinine_phosphokinase  299 non-null    int64  
 3   diabetes                  299 non-null    int64  
 4   ejection_fraction         299 non-null    int64  
 5   high_blood_pressure       299 non-null    int64  
 6   platelets                 299 non-null    float64
 7   serum_creatinine          299 non-null    float64
 8   serum_sodium              299 non-null    int64  
 9   sex                       299 non-null    int64  
 10  smoking                   299 non-null    int64  
 11  time                      299 non-null    int64  
 12  DEATH_EVENT               299 non-null    int64  
dtypes: float64(3), int64(10)
memory usage: 30.5 KB


# Correlations

In [545]:
corr_matrix = clinical_records.corr()
corr_matrix["DEATH_EVENT"].sort_values(ascending=False)

DEATH_EVENT                 1.000000
serum_creatinine            0.294278
age                         0.253729
high_blood_pressure         0.079351
anaemia                     0.066270
creatinine_phosphokinase    0.062728
diabetes                   -0.001943
sex                        -0.004316
smoking                    -0.012623
platelets                  -0.049139
serum_sodium               -0.195204
ejection_fraction          -0.268603
time                       -0.526964
Name: DEATH_EVENT, dtype: float64

In [546]:
y = clinical_records["DEATH_EVENT"]

In [547]:
new_x =clinical_records[['age', 'ejection_fraction', 'serum_sodium', 'time']]

In [548]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(new_x, y, test_size = 0.2, random_state = 2698)

In [549]:
X_train

age  ejection_fraction  serum_sodium  time
282  42.0                 30           128   250
246  55.0                 25           138   214
58   49.0                 20           136    55
84   59.0                 25           141    78
233  53.0                 38           137   209
..    ...                ...           ...   ...
199  60.0                 35           113   186
195  77.0                 45           145   180
280  70.0                 40           137   247
9    80.0                 35           133    10
120  60.0                 60           135    95

[239 rows x 4 columns]

# Models

# Logistic Regression

In [550]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression(max_iter = 1000)
log_reg.fit(X_train, y_train)
log_pred = log_reg.predict(X_test)
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, log_pred))

0.9


# Random Forest

In [551]:
from sklearn.ensemble import RandomForestClassifier

forest_clf = RandomForestClassifier(random_state=5)

forest_clf.fit(X_train, y_train)

forest_pred = forest_clf.predict(X_test)

accuracy_score(y_test, forest_pred)

0.95

# Gradient Boosting

In [552]:
from sklearn.ensemble import GradientBoostingClassifier
gradientboost_clf = GradientBoostingClassifier(max_depth = 1)
gradientboost_clf.fit(X_train,y_train)
gradientboost_pred = gradientboost_clf.predict(X_test)
accuracy_score(y_test,gradientboost_pred)

0.9

# SVM

In [553]:
from sklearn.svm import SVC
svm = SVC()
svm.fit(X_train,y_train)
svc_pred = svm.predict(X_test)
accuracy_score(y_test,svc_pred)

0.9166666666666666

# AdaBoost

In [554]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

In [555]:
ada_clf = AdaBoostClassifier(
 DecisionTreeClassifier(max_depth=1), n_estimators=200,
 algorithm="SAMME.R", learning_rate=0.3)
ada_clf.fit(X_train, y_train)

ada_pred = ada_clf.predict(X_test)

accuracy_score(y_test, ada_pred)

0.9333333333333333

# Final Results:

Logistic Regression: 90%

Random Forest: 95%

Gradient Boost: 90%

SVM: 90%

AdaBoost: 93.3%

# Bonus Tip:

If you change the random_state paramater when performing train_test_split, you actually get very different results. For example, when you change the random_state to random_state=1, the results are as follows:

Logistic Regression: 86.6%

Random Forest: 86.6%

Gradient Boost: 83.3%

SVM: 90%

AdaBoost: 85%

As you can see, when we change the random_state parameter, we get very different results which lead to worse performing models.

Another example, that really takes a toll on the accuracies is when I change the random_state paramater to random_state=42, which is what many practioners use as the random_state. The results are as follows:

Logistic Regression: 81.6%

Random Forest: 76.6%

Gradient Boost: 78.3%

SVM: 71.6%

AdaBoost: 71.6%

These results are not very appealing and this was only due to changing the random_state paramter. We don't get anywhere near the results that we did when random_state was equal to 1 or 2 and this was something I wanted to share and a cool discovery I had. I didn't realize that changing such a simple paramter could really be the difference between the highest accuracy of 93.3% and 81.6%. Hope you learnt something.